src/charm.py

```diff
@@ -8,18 +8,223 @@
+            # Push the config on first layer update to avoid PathError
+            self._container.push(self._hydra_config_path, self._config, make_dirs=True)
+            logger.info("Pushed hydra config")
+
+        try:
+            current_config = self._container.pull(self._hydra_config_path).read()
+        except (ProtocolError, PathError) as err:
+            logger.error(str(err))
+            self.unit.status = BlockedStatus(str(err))
+        else:
+            if current_config != self._config:
+                self._container.push(self._hydra_config_path, self._config, make_dirs=True)
+                logger.info("Updated hydra config")
+
+        try:
+            logger.info("Replanning pebble container")
+            self._container.replan()
+        except ChangeError as err:
+            logger.error(str(err))
+            self.unit.status = BlockedStatus("Failed to replan")
+            return
+
+    def _update_container(self, event) -> None:
+        """Update configs, pebble layer and run database migration."""
+        try:
+            self._check_leader()
+        except CheckFailedError as err:
+            self.model.unit.status = err.status
+            return
+
+        if not self.model.relations["pg-database"]:
+            logger.error("Missing required relation with postgresql")
+            self.model.unit.status = BlockedStatus("Missing required relation with postgresql")
+            return
+
+        if not self.database.is_database_created():
+            event.defer()
+            logger.info("Missing database details. Deferring the event.")
+            self.unit.status = WaitingStatus("Waiting for database creation")
+            return
+
+        if not self._container.can_connect():
+            event.defer()
+            logger.info("Cannot connect to Hydra container. Deferring the event.")
+            self.unit.status = WaitingStatus("Waiting to connect to Hydra container")
+            return
+
+        self._update_layer()
+
+        self._run_sql_migration()
+
+    def _run_sql_migration(self) -> None:
+        """Runs a command to create SQL schemas and apply migration plans."""
+        process = self._container.exec(
+            ["hydra", "migrate", "sql", "-e", "--config", self._hydra_config_path, "--yes"],
+            timeout=20.0,
+        )
+        try:
+            stdout, _ = process.wait_output()
+            logger.info(f"Executing automigration: {stdout}")
+        except ExecError as err:
+            logger.error(f"Exited with code {err.exit_code}. Stderr: {err.stderr}")
+            self.unit.status = BlockedStatus("Database migration job failed")
+
+    def _on_hydra_pebble_ready(self, event) -> None:
+        """Event Handler for pebble ready event."""
+        self.model.unit.status = MaintenanceStatus("Configuring hydra container")
+        self._update_container(event)
+
+        self.model.unit.status = ActiveStatus()
```

Comment:
You can't set active status here. If container cannot connect/missing database relation/waiting for database creation, this active status will overwrite the blocked or waiting status set in `_update_container`. Same in `_on_db_events`